Content.Server/Atmos/Portable/SpaceHeaterSystem.cs

```diff
@@ -0,0 +1,198 @@
+using Content.Server.Power.EntitySystems;
+using Content.Shared.UserInterface;
+using Robust.Server.GameObjects;
+
+namespace Content.Server.Atmos.Portable;
+
+public sealed class SpaceHeaterSystem : EntitySystem
+{
+    [Dependency] private readonly AtmosphereSystem _atmosphereSystem = default!;
+    [Dependency] private readonly PopupSystem _popup = default!;
+    [Dependency] private readonly PowerReceiverSystem _power = default!;
+    [Dependency] private readonly SharedAppearanceSystem _appearance = default!;
+    [Dependency] private readonly UserInterfaceSystem _userInterfaceSystem = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<SpaceHeaterComponent, ActivatableUIOpenAttemptEvent>(OnUIActivationAttempt);
+        SubscribeLocalEvent<SpaceHeaterComponent, BeforeActivatableUIOpenEvent>(OnBeforeOpened);
+
+        SubscribeLocalEvent<SpaceHeaterComponent, AtmosDeviceUpdateEvent>(OnDeviceUpdated);
+        SubscribeLocalEvent<SpaceHeaterComponent, MapInitEvent>(OnInit);
+        SubscribeLocalEvent<SpaceHeaterComponent, PowerChangedEvent>(OnPowerChanged);
+
+        SubscribeLocalEvent<SpaceHeaterComponent, SpaceHeaterToggleMessage>(OnToggle);
+        SubscribeLocalEvent<SpaceHeaterComponent, SpaceHeaterChangeTemperatureMessage>(OnTemperatureChanged);
+        SubscribeLocalEvent<SpaceHeaterComponent, SpaceHeaterChangeModeMessage>(OnModeChanged);
+        SubscribeLocalEvent<SpaceHeaterComponent, SpaceHeaterChangePowerLevelMessage>(OnPowerLevelChanged);
+    }
+
+    private void OnInit(EntityUid uid, SpaceHeaterComponent spaceHeater, MapInitEvent args)
+    {
+        if (!TryComp<GasThermoMachineComponent>(uid, out var thermoMachine))
+            return;
+        thermoMachine.Cp = spaceHeater.HeatingCp;
+        thermoMachine.HeatCapacity = spaceHeater.PowerConsumption;
+    }
+
+    private void OnBeforeOpened(EntityUid uid, SpaceHeaterComponent spaceHeater, BeforeActivatableUIOpenEvent args)
+    {
+        DirtyUI(uid, spaceHeater);
+    }
+
+    private void OnUIActivationAttempt(EntityUid uid, SpaceHeaterComponent spaceHeater, ActivatableUIOpenAttemptEvent args)
+    {
+        if (!Comp<TransformComponent>(uid).Anchored)
+        {
+            _popup.PopupEntity(Loc.GetString("comp-space-heater-unanchored", ("device", Loc.GetString("comp-space-heater-device-name"))), uid, args.User);
+            args.Cancel();
+        }
+    }
+
+    private void OnDeviceUpdated(EntityUid uid, SpaceHeaterComponent spaceHeater, ref AtmosDeviceUpdateEvent args)
+    {
+        if (!_power.IsPowered(uid)
+            || !TryComp<GasThermoMachineComponent>(uid, out var thermoMachine))
+        {
+            return;
+        }
+
+        // First get the heat direction of the thermomachine (if any) and update appeareance accordingly
+        if (thermoMachine.LastEnergyDelta > 0)
+        {
+            spaceHeater.State = SpaceHeaterState.Heating;
```

Comment:
`spaceHeater.State` seems to duplicate state that is stored elsewhere. This makes it confusing to understand and maintain, e.g. who is responsible for setting `spaceHeater.State` when something changes?

For example, it seems to me that `spaceHeater.State` is used only in `UpdateApperance()`. If that is the case, then this block computing what the state should be moved to `UpdateApperance()` and the extra calls to set it removed.

Reply:
You're right, that's a leftover of an earlier attempt and the space heater state isn't used for anything besides the visualizer. It's gone now.